File: apps/query-burst-msa/order-service/src/main/java/com/booster/queryburstmsa/order/web/OrderController.java
package com.booster.queryburstmsa.order.web;

import com.booster.queryburstmsa.order.application.OrderApplicationService;
import com.booster.queryburstmsa.order.web.dto.OrderCreateRequest;
import com.booster.queryburstmsa.order.web.dto.OrderResponse;
import com.booster.queryburstmsa.order.web.dto.OrderSummaryResponse;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.net.URI;
import java.util.List;

@RestController
@RequestMapping("/api/orders")
public class OrderController {

    private final OrderApplicationService orderApplicationService;

    public OrderController(OrderApplicationService orderApplicationService) {
        this.orderApplicationService = orderApplicationService;
    }

    @GetMapping
    public ResponseEntity<List<OrderSummaryResponse>> getOrders(
            @RequestParam(required = false) Long cursor,
            @RequestParam(defaultValue = "20") int size
    ) {
        return ResponseEntity.ok(orderApplicationService.getOrders(cursor, size));
    }

    @GetMapping("/{orderId}")
    public ResponseEntity<OrderResponse> getOrder(@PathVariable Long orderId) {
        OrderResponse response = orderApplicationService.getOrder(orderId);
        return response == null ? ResponseEntity.notFound().build() : ResponseEntity.ok(response);
    }

    @PostMapping
    public ResponseEntity<OrderResponse> createOrder(
            @RequestBody OrderCreateRequest request,
            @RequestHeader(value = "Idempotency-Key", required = false) String idempotencyKey
    ) {
        OrderResponse created = orderApplicationService.createOrder(request, idempotencyKey);
        return ResponseEntity.created(URI.create("/api/orders/" + created.orderId())).body(created);
    }

    @PatchMapping("/{orderId}/pay")
    public ResponseEntity<Void> pay(@PathVariable Long orderId) {
        orderApplicationService.pay(orderId);
        return ResponseEntity.noContent().build();
    }

    @PatchMapping("/{orderId}/ship")
    public ResponseEntity<Void> ship(@PathVariable Long orderId) {
        orderApplicationService.ship(orderId);
        return ResponseEntity.noContent().build();
    }

    @PatchMapping("/{orderId}/deliver")
    public ResponseEntity<Void> deliver(@PathVariable Long orderId) {
        orderApplicationService.deliver(orderId);
        return ResponseEntity.noContent().build();
    }

    @DeleteMapping("/{orderId}")
    public ResponseEntity<Void> cancel(@PathVariable Long orderId) {
        orderApplicationService.cancel(orderId);
        return ResponseEntity.noContent().build();
    }
}

File: apps/kotlin/shopping-service/src/main/kotlin/com/booster/kotlin/shoppingservice/order/web/controller/OrderController.kt
package com.booster.kotlin.shoppingservice.order.web.controller

import com.booster.kotlin.shoppingservice.common.response.ApiResponse
import com.booster.kotlin.shoppingservice.order.application.OrderService
import com.booster.kotlin.shoppingservice.order.application.dto.CancelOrderCommand
import com.booster.kotlin.shoppingservice.order.web.dto.request.CreateOrderRequest
import com.booster.kotlin.shoppingservice.order.web.dto.response.OrderResponse
import com.booster.kotlin.shoppingservice.order.web.dto.response.OrderSummaryResponse
import jakarta.validation.Valid
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.web.PageableDefault
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/v1/orders")
class OrderController(
    private val orderService: OrderService,
) {

    @GetMapping
    fun getOrders(
        @AuthenticationPrincipal userId: Long,
        @PageableDefault(size = 10, sort = ["createdAt"], direction = Sort.Direction.DESC) pageable: Pageable,
    ): ResponseEntity<ApiResponse<Page<OrderSummaryResponse>>> {
        val orders = orderService.getOrders(userId, pageable).map { OrderSummaryResponse.from(it) }
        return ResponseEntity.ok(ApiResponse.ok(orders))
    }

    @GetMapping("/{orderId}")
    fun getOrder(
        @AuthenticationPrincipal userId: Long,
        @PathVariable orderId: Long,
    ): ResponseEntity<ApiResponse<OrderResponse>> {
        val order = orderService.getOrder(userId, orderId)
        return ResponseEntity.ok(ApiResponse.ok(OrderResponse.from(order)))
    }

    @PostMapping
    fun create(
        @AuthenticationPrincipal userId: Long,
        @RequestBody @Valid request: CreateOrderRequest,
    ): ResponseEntity<ApiResponse<OrderResponse>> {
        val order = orderService.create(request.toCommand(userId))
        return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.ok(OrderResponse.from(order)))
    }

    @PostMapping("/{orderId}/cancel")
    fun cancel(
        @AuthenticationPrincipal userId: Long,
        @PathVariable orderId: Long,
        @RequestParam(required = false) reason: String?,
    ): ResponseEntity<ApiResponse<OrderResponse>> {
        val order = orderService.cancel(CancelOrderCommand(userId, orderId, reason))
        return ResponseEntity.ok(ApiResponse.ok(OrderResponse.from(order)))
    }
}
